src/Build/BackEnd/Components/BuildRequestEngine/BuildRequestEngine.cs

```diff
@@ -1137,7 +1137,7 @@ private void IssueBuildRequests(BuildRequestEntry issuingEntry, FullyQualifiedBu
                             configurationId: request.Config.ConfigurationId,
                             escapedTargets: request.Targets,
                             hostServices: issuingEntry.Request.HostServices,
-                            parentBuildEventContext: issuingEntry.Request.BuildEventContext,
+                            parentBuildEventContext: issuingEntry.Request.CurrentMSBuildTask ?? issuingEntry.Request.BuildEventContext,
```

Comment:
I think `CurrentMSBuildTask` is written by a different thread than the one reading it here. Should it be volatile to flush the value to memory, or am I just too paranoid? `BuildEventContext` should be fine since it gets set very early during the request creation.

Reply:
Yes, it's written by the RequestBuilder thread, and read in BuildRequestEngine (dataflow callback ProcessMessage). However while it's read the RequestBuilder thread is not running, it's awaiting task completion so there's no parallelism.

Reply:
I was thinking about the case where the value is not yet available in the reading thread's core's cache. AFAIK `volatile` (and at least in Java, other synchronization primitives) flush the writes from the cache, and force the reads from memory (https://igoro.com/archive/volatile-keyword-in-c-memory-model-explained/)

Reply:
this is too deep for me ;)

Reply:
I looked into it a bit more, and I think it's actually fine. The read is inside a lock block, which refreshes the cache on entry. And the write is inside an async method, which apparently also does it: https://stackoverflow.com/q/6581848/417199

Safest is to make `BuildRequest.CurrentMSBuildTask ` volatile, but I doubt anybody will remove the lock or the async.